### PR TITLE
[MRG] NetworkPlot class to enable interactive visualizations and animations

### PR DESCRIPTION
**Pull request by ntolley**
Here's my first pass at creating a viz class that can be used to quickly render the full 3D network, as well as membrane potentials of individual sections. This is meant to be largely generic and flexible so that it can be co-opted for other purposes. I think the most interesting applications at the moment would be:

- Quickly generating animations of HNN simulations
- Embedding the 3D plot into the current ipywidgets GUI

The current strategy for the class is to do all of the heavy lifting at the initialization (namely creating the 3D object, and mapping the voltage recordings to the appropriate colors). For a quick demo, the code below can be run on my laptop on a single core in approximately 20 sec:
```py
from hnn_core import jones_2009_model, simulate_dipole
from hnn_core.network_models import add_erp_drives_to_jones_model
from hnn_core.viz import NetworkPlot

net = jones_2009_model()
net.set_cell_positions(inplane_distance=300)
add_erp_drives_to_jones_model(net)
dpl = simulate_dipole(net, dt=0.5, tstop=170, record_vsec='all')

net_plot = NetworkPlot(net)
```
image

If you want to play around with visualizing the voltages, you can try the following code block:
```py
from ipywidgets import interact, IntSlider

def update_plot(t_idx, elev, azim):
    net_plot.update_section_voltages(t_idx)
    net_plot.elev = elev
    net_plot.azim = azim
    return net_plot.fig

time_slider = IntSlider(min=0, max=len(net_plot.times), value=1, continuous_update=False)
elev_slider = IntSlider(min=-100,max=100, value=10, continuous_update=False)
azim_slider = IntSlider(min=-100,max=100, value=-100, continuous_update=False)

interact(update_plot, t_idx=time_slider, elev=elev_slider, azim=azim_slider)
```
image

### CODE REVIEW

**Comment by ntolley** [comment]
@jasmainak @rythorpe perhaps it'd be a good idea to make this a separate file? If we can find a solution to updating the section colors very rapidly, we can even consider adding a "play" button inside the GUI (or on the HNN website!) like the example here: https://matplotlib.org/stable/gallery/animation/unchained.html

@chenghuzi you might have some good insight on how to make the current code more "reactive"

File: examples/howto/plot_hnn_animation.py
Comment (jasmainak) [inline examples/howto/plot_hnn_animation.py:3]: to be fixed :)

File: hnn_core/viz.py
Comment (jasmainak) [inline hnn_core/viz.py:1428]: an `export` method? I think matplotlib allows exporting animations ... I want to see HNN animations in presentations of people I don't know ;-)

**Comment by jasmainak** [comment]
Agnostic about saving to a different file ... see my comment about the azimuth/elevation control. Maybe it makes the class thinner?

File: hnn_core/viz.py
Comment (jasmainak) [inline hnn_core/viz.py:1203]:
```suggestion
class NetworkDataPlotter:
```

?

Comment (ntolley) [reply hnn_core/viz.py:1203]: `NetworkPlotter`? It might useful too for just visualizing the 3D morphology without any data too

**Comment by ntolley** [comment]
Dropping this here for easy testing once we get closer to merging. I think rendering movies is going to be a longish wait for larger networks, but not absurdly so (maybe like 5-10 minutes? if you downsample?)
```py
import hnn_core
import os.path as op
from hnn_core import jones_2009_model, simulate_dipole, read_params
from hnn_core.network_models import add_erp_drives_to_jones_model
from hnn_core.viz import NetworkPlot

hnn_core_root = op.dirname(hnn_core.__file__)
params_fname = op.join(hnn_core_root, 'param', 'default.json')
params = read_params(params_fname)
params.update({'N_pyr_x': 3, 'N_pyr_y': 3})
net = jones_2009_model(params)
net.set_cell_positions(inplane_distance=300)
add_erp_drives_to_jones_model(net)
dpl = simulate_dipole(net, dt=0.5, tstop=170, record_vsec='all')

net_plot = NetworkPlot(net)
net_plot.export_movie('demo.gif', dpi=200)
```
demo

File: hnn_core/viz.py
Comment (ntolley) [inline hnn_core/viz.py:1325]: @jasmainak here's the export function!

Comment (ntolley) [reply hnn_core/viz.py:1325]: There are some solutions here that might speed up the exporting by circumventing the matplotlib animation module:
https://stackoverflow.com/questions/4092927/generating-movie-from-python-without-saving-individual-frames-to-files

File: hnn_core/viz.py
Comment (jasmainak) [reply hnn_core/viz.py:1203]: Yeah if data is None

File: hnn_core/viz.py
Comment (jasmainak) [reply hnn_core/viz.py:1325]: Is it possible to add a progressbar using a callback? I don't think the speed is a showstopper here. Maybe a param called "decim" could be added to take every nth frame. Make it work, make it nice, make it fast ...

**Comment by chenghuzi** [comment]
> @jasmainak @rythorpe perhaps it'd be a good idea to make this a separate file? If we can find a solution to updating the section colors very rapidly, we can even consider adding a "play" button inside the GUI (or on the HNN website!) like the example here: https://matplotlib.org/stable/gallery/animation/unchained.html
> 
> @chenghuzi you might have some good insight on how to make the current code more "reactive"

Great work! Although, I'm concerned about the reactivity of our GUI. We are taking snapshots from the buffer and using ﻿`imshow` to render network images for now. So, it might be difficult to have real-time interaction in the new GUI with these new settings. But this will not affect the current function so there is nothing else to worry about.

File: hnn_core/viz.py
Comment (ntolley) [reply hnn_core/viz.py]: Definitely would be a nice addition but I think I'd need to be much more matplotlib savvy to know how to implement that :smile:

File: hnn_core/viz.py
Comment (rythorpe) [inline hnn_core/viz.py:1319]: It'd be nice if we could populate `self.color_array` with zeros or ones at the onset so that the network can be plotted even if the simulation is run with `record_vsec=False`.

Comment (ntolley) [reply hnn_core/viz.py:1319]: Super good suggestion! Also makes for a good smoke test to plot a simulated/unsimulated network

Comment (rythorpe) [reply hnn_core/viz.py:1319]: Overall, this branch is such a cool feature! Super excited to get it merged.

File: examples/howto/plot_hnn_animation.py
Comment (rythorpe) [inline examples/howto/plot_hnn_animation.py:15]:
```suggestion
from hnn_core.viz import NetworkPlotter
```

Comment (rythorpe) [reply examples/howto/plot_hnn_animation.py:15]: I'm guessing the name of this class got changed somewhere along the way and you forgot to update it here?

File: examples/howto/plot_hnn_animation.py
Comment (rythorpe) [inline examples/howto/plot_hnn_animation.py:33]: For some reason the interactive slider didn't work for me. It might be a version issue and totally unrelated to your code here. Have you encountered a similar issue at any point @ntolley?

**Comment by ntolley** [comment]
@rythorpe @jasmainak this PR is ready for review! Two weird errors have come up however. First it seems like sphinx-gallery doesn't support python versions older than 3.8
image

Does this mean our minimum python version needs to be updated?

Also the windows unit tests are failing with an error in the GUI code...
image

This is the offending line in case you guys have any ideas on the possible cause:
https://github.com/ntolley/hnn-core/blob/e3aa27c91ba4ed7478d2214e53fea618b793f45b/hnn_core/tests/test_viz.py#L23C5-L23C5

**Comment by ntolley** [comment]
Yup just confirmed, sphinx now requires python>=3.8

https://github.com/sphinx-gallery/sphinx-gallery/blob/e2903c9e43572a0a29850f7ab153ff69098f5c81/setup.py#L58C5-L58C5

File: .circleci/config.yml
Comment (rythorpe) [inline .circleci/config.yml:30]: Do you think we need to update `setup.py` and our unit tests with `python>=3.8` as well to ensure all builds are getting tested properly?

**Comment by rythorpe** [comment]
> @rythorpe @jasmainak this PR is ready for review! Two weird errors have come up however. First it seems like sphinx-gallery doesn't support python versions older than 3.8 image
> 
> Does this mean our minimum python version needs to be updated?
> 
> Also the windows unit tests are failing with an error in the GUI code... image
> 
> This is the offending line in case you guys have any ideas on the possible cause: https://github.com/ntolley/hnn-core/blob/e3aa27c91ba4ed7478d2214e53fea618b793f45b/hnn_core/tests/test_viz.py#L23C5-L23C5

Ugh, I'm guessing the most recent version of matplotlib has diverged for windows vs. linux. Something to do with the `agg` backend....

**Comment by ntolley** [comment]
Ok updating the python version used for CircleCI worked, however it seems something in the MNE code is breaking the somato example now:
image

File: .circleci/config.yml
Comment (jasmainak) [reply .circleci/config.yml:30]: that's true but you won't realize if other dependencies start requiring 3.8 now that you have updated it here. I don't think we have a CI that tests for python < 3.8? Better to be consistent and update readme/setup etc.

File: examples/howto/plot_hnn_animation.py
Comment (jasmainak) [inline examples/howto/plot_hnn_animation.py:16]:
```suggestion
import os.path as op

import hnn_core
```

File: examples/howto/plot_hnn_animation.py
Comment (jasmainak) [inline examples/howto/plot_hnn_animation.py:26]: this won't work with the `hdf5` format. We'll need an API to do this.

File: examples/howto/plot_hnn_animation.py
Comment (jasmainak) [inline examples/howto/plot_hnn_animation.py:39]: I don't see a figure in `ipython`. Maybe you need a `plt.show` or `plt.draw` ? Maybe `net_plot.show()` ?

File: hnn_core/viz.py
Comment (jasmainak) [inline hnn_core/viz.py:1242]: needs to be single line according to pep 257

File: hnn_core/viz.py
Comment (jasmainak) [inline hnn_core/viz.py:1259]: we're missing a `colorbar` parameter that would default to `True`

File: hnn_core/viz.py
Comment (jasmainak) [inline hnn_core/viz.py:1395]: needs a `Parameters` heading and an empty line

File: examples/howto/plot_hnn_animation.py
Comment (jasmainak) [inline examples/howto/plot_hnn_animation.py:56]: we can leave this for now, but in the future I'd be in favor of deprecating this function for a boilerplate file that contains the drives in an `hdf5` file and we read it directly from that ... functions should not be black boxes ...

File: hnn_core/tests/test_viz.py
Comment (jasmainak) [inline hnn_core/tests/test_viz.py:238]: I guess github co-pilot wasn't clever enough to compress this in a loop? ;-)

File: hnn_core/tests/test_viz.py
Comment (jasmainak) [inline hnn_core/tests/test_viz.py:290]: why is this repeated?

File: examples/howto/plot_hnn_animation.py
Comment (jasmainak) [inline examples/howto/plot_hnn_animation.py:67]: something is wrong with the default parameters ... it doesn't play well with our default simulation parameters. I tried:

```py
net_plot.export_movie('test.mov')
```

and it was stuck forever.

File: hnn_core/viz.py
Comment (jasmainak) [inline hnn_core/viz.py:1240]: needs to be added to `api.rst`

File: .circleci/config.yml
Comment (rythorpe) [reply .circleci/config.yml:30]: I went ahead and updated this in #677. We'll see if it fixes the Windows `test_viz.py` error....

Comment (jasmainak) [reply .circleci/config.yml:30]: excellent!!

**Comment by jasmainak** [comment]
> it seems something in the MNE code is breaking the somato example now:

looks like `nibabel` is now being used in MNE for reading the surface files ... I reported this upstream:

https://github.com/mne-tools/mne-python/pull/11565/files#r1337642633

hopefully it'll get fixed soon

File: hnn_core/tests/test_viz.py
Comment (ntolley) [reply hnn_core/tests/test_viz.py:290]: It is redundant, but I figure there should be separate tests for setting parameters before or after initialization

File: examples/howto/plot_hnn_animation.py
Comment (ntolley) [reply examples/howto/plot_hnn_animation.py:26]: are you ok to leave this for now and replace when we push #651 over the finish line?

File: examples/howto/plot_hnn_animation.py
Comment (ntolley) [reply examples/howto/plot_hnn_animation.py:40]: This shows up in circleCI docs: 
image

honestly this issue of when we do and don't need to use `plt.show()` is really opaque to me, it is only a problem sometimes on some certain systems...

File: examples/howto/plot_hnn_animation.py
Comment (ntolley) [reply examples/howto/plot_hnn_animation.py:68]: how long did you wait? on my laptop it takes 1min 30sec to generate the gif

The default `writer` is `pillow` which doesn't support the `.mov` extension. I was able to export using the `ffmpeg` writer.

Having a programmatic check on the extensions would be nice to have, but it seems like it would be a bit annoying to query this for every valid writer: https://stackoverflow.com/questions/22571044/what-formats-can-matplotlib-animations-be-saved-as 

(I've dug into the attributes and unfortunately there isn't a clear list I can find)

**Comment by jasmainak** [comment]
@rythorpe @gtdang could you guys review and merge this PR if it's good by you? I am finding it difficult to find time and don't want to block progress ... this could be a good excuse for a release.

File: examples/howto/plot_hnn_animation.py
Comment (rythorpe) [reply examples/howto/plot_hnn_animation.py:27]: I think we can now remove the reference to `default.json` and avoiding using/modifying the params dict here.

File: hnn_core/cell.py
Comment (rythorpe) [inline hnn_core/cell.py:754]: These new args should be added to the docstring.

File: hnn_core/viz.py
Comment (rythorpe) [inline hnn_core/viz.py:852]: Same here: add to docstring.

**Review by rythorpe — COMMENTED** [review]
This is almost ready @ntolley! I just tested it out and made a few minor comments.

File: hnn_core/tests/test_viz.py
Comment (rythorpe) [reply hnn_core/tests/test_viz.py:298]: This is a nitpick, but could we put these tests in `for` loop to conserve space + enhance readability?

**Comment by rythorpe** [comment]
Just in case this slipped your attention @ntolley, this PR is sooooo close ;-)

File: hnn_core/viz.py
Comment (gtdang) [inline hnn_core/viz.py:914]: It doesn't seem like these lines need to be nested in this for loop? Seems like all of those values are static. Should they be using pt instead of pos? Or the iterated section?

File: hnn_core/viz.py
Comment (gtdang) [inline hnn_core/viz.py:1299]: I would suggest moving all the type validation to the top of the init for both logic and readability. I'm assuming that if any of the validations fail an error is being raised and you won't continue with the init?

**Review by gtdang — CHANGES_REQUESTED** [review]

File: hnn_core/tests/test_viz.py
Comment (gtdang) [inline hnn_core/tests/test_viz.py:225]: The tests in this module are too long. Generally, each unit test should focus on one specific aspect of the behavior of a function to better isolate location of bugs. That refactor should probably come in a separate PR, but for this one I think the NetworkPlotter should be spun out into its own test at the very least. Ideally it would be broken even further into the main operations you are testing. Here are the separate operations I see from a quick glance: 

1. init  keyword validation 
2. setter validation 
3. simulation validation 
4. movie export 

Creating a pytest fixture for the network will help with redundancies. Also 2nd creating a loop for the pytest raises. You can use the f-strings and dictionary unpacking. Something like:
```python
args = ['xlim', 'ylim', 'zlim']
    for arg in args:
        d_kw = {arg: 'blah'}
        with pytest.raises(TypeError, match=f'{arg} must be'):
            _ = NetworkPlotter(net, **d_kw)
```

File: hnn_core/tests/test_viz.py
Comment (gtdang) [inline hnn_core/tests/test_viz.py:336]: I guess this will test if this operation crashes. A more comprehensive test would be to make sure this file actually exists. By following up with an os.path.exists() or os.path.isfile(). Bonus points for writing to temp space.

**Review by gtdang — CHANGES_REQUESTED** [review]

File: hnn_core/tests/test_viz.py
Comment (ntolley) [reply hnn_core/tests/test_viz.py:235]: Thanks for the suggestions @gtdang! Just updated the tests accordingly, let me know what you think

**Comment by rythorpe** [comment]
@ntolley do you know if `plot_cell_morphology` sets the length of each cell section before plotting it? Or are the section endpoints already updated with the appropriate length? Apologies if we already hashed this out, I just can't remember and was wondering about this the other day....

**Comment by ntolley** [comment]
@rythorpe they are updated automatically! #433 fixed this early on, but then Rajat's project made a new and improved version where he implemented the Neuron functions for calculating end points directly in our python code (necessary for saving this infromation in a JSON/dict format)

File: hnn_core/tests/test_viz.py
Comment (ntolley) [reply hnn_core/tests/test_viz.py]: Borrowed your logic from the network config PR FOR THIS ONE

File: hnn_core/tests/test_viz.py
Comment (gtdang) [inline hnn_core/tests/test_viz.py:255]: If you wanted a loop for these you could do it like this:

```python
not_num = 'string'
not_int = 1.0
test_kws = {'xlim': not_num, 'ylim': not_num,
            'trial_idx': not_int, 'time_idx': not_int,
            }
for keyword, value in test_kws.items():
    kwargs = {keyword: value}
    with pytest.raises(TypeError, match=f'{keyword} must be'):
        _ = NetworkPlotter(net, **kwargs)
```

But if up to you if you find it less interpretable!

**Review by gtdang — COMMENTED** [review]
Looking good! Like the test update and the init. Left one comment on the tests.

File: hnn_core/tests/test_viz.py
Comment (gtdang) [inline hnn_core/tests/test_viz.py:240]: Ah that's simpler! Nice!

**Comment by ntolley** [comment]
@jasmainak this is ready for a final round of reviews when you have the chance!

I think there are some tasks that might be delegated to followup PR's (e.g. a better jupyter notebook tutorial?), but in terms of basic functionality and readability I think this is in a good place

File: hnn_core/viz.py
Comment (jasmainak) [inline hnn_core/viz.py:1475]: use the docstring for comments :)

File: hnn_core/viz.py
Comment (jasmainak) [inline hnn_core/viz.py:1506]: our codespell checker isn't catching this?

**Comment by jasmainak** [comment]
@ntolley this is tremendous!!! 🥳 when can I have the privilege of merging it? I tried it and it works really nice. @rythorpe did you want to look? It's good to go on my end.

One feature request: perhaps we can have a small text in the corner showing the timestamp ... I was a bit confused whether the plot was still updating (or is it instantaneous?) when I moved it around.

**Comment by jasmainak** [comment]
@gtdang do you still have remaining objections? Merging seems blocked because of your comments

**Review by jasmainak — APPROVED** [review]

**Comment by jasmainak** [comment]
Don't forget to update `whats_new.rst` !

**Review by gtdang — APPROVED** [review]
Looks good!

File: doc/whats_new.rst
Comment (rythorpe) [inline doc/whats_new.rst:62]:
```suggestion
  by `Nick Tolley`_ in :gh:`649`.
```
I know we haven't been good about this so far, but we should try to be more consisent with punctuation.

**Comment by rythorpe** [comment]
Aside from @jasmainak's spelling error and my incredibly nitpicky punctuation suggestion above, this looks great @ntolley !

File: hnn_core/viz.py
Comment (ntolley) [inline hnn_core/viz.py:1506]:
```suggestion
    # Elevation and azimuth of 3D viewpoint
```

**Comment by rythorpe** [comment]
I've enabled auto-merge! Thanks @ntolley for this incredible feature - I'm very excited to put it to good use :smile:

**Pull request merged**
